Migrate CacheStorageConnection::open to NativePromise

youennf opened the pull request:
#### 43bf0bb142ab5ea9b74e0cd70cbca969aacb8840
<pre>
Migrate CacheStorageConnection::open to NativePromise
<a href="https://bugs.webkit.org/show_bug.cgi?id=266568">https://bugs.webkit.org/show_bug.cgi?id=266568</a>
<a href="https://rdar.apple.com/119804320">rdar://119804320</a>

Reviewed by Jean-Yves Avenard.

The goal is first to simplify CacheStorage operations handling from workers.
We should as a follow-up step allow to send/receive IPC out of main thread, which will be made easier with promises.

Given it runs in workers, we need to have a way to register NativePromise callbacks that are guaranteed to be destroyed in workers.
We do this by registering these callbacks in a WorkerGlobalScope so that we can destroy them when WorkerGlobalScope gets stopped.

To do this, we allow a promise to either take a RefCountedGuaranteedSerialFunctionDispatcher or a NativePromiseDispatcher.
The latter is a RefCountedSerialFunctionDispatcher but without any guarantee that the function will be dispatched or not.
To handle this possibility, the NativePromiseDispatcher is responsible to disconnect the NativePromiseCallback that it is given in registerCallback
in the same context as its RefCountedSerialFunctionDispatcher.

* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/NativePromise.h:
(WTF::NativePromiseRequest::~NativePromiseRequest):
(WTF::NativePromiseRequest::track):
(WTF::NativePromiseRequest::operator bool const):
(WTF::NativePromiseRequest::complete):
(WTF::NativePromiseRequest::disconnect):
* Source/WebCore/Modules/cache/CacheStorageConnection.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::DOMCacheStorage::doOpen):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::open):
(WebCore::WorkerCacheStorageConnection::clearPendingRequests):
(WebCore::WorkerCacheStorageConnection::openCompleted): Deleted.
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::stopActiveDOMObjects):
(WebCore::ScriptExecutionContext::nativePromiseDispatcher):
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::Task::Task):
(WebCore::ScriptExecutionContext::Task::m_isCleanupTask):
(WebCore::ScriptExecutionContext::enqueueTaskWhenPromiseIsSettled):
(WebCore::ScriptExecutionContext::AddConsoleMessageTask::AddConsoleMessageTask):
* Source/WebCore/page/CacheStorageProvider.h:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::open):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/273828@main">https://commits.webkit.org/273828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3cbc688a1f27f9a5213e51038b126d403ce5556

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/36953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39445 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32965 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38049 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12854 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31524 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37366 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13285 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32525 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11608 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32813 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40694 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33384 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33155 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37534 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/36546 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11898 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9714 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35654 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13557 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/43364 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8341 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12291 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/8959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12798 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -312,6 +312,34 @@ struct VoidPlaceholder {
 };
 } // namespace detail
 
+// Request is the object returned by NativePromise::then()/whenSettled is used by NativePromiseRequest holder to track/disconnect.
```
Obsolete comment ?

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -312,6 +312,34 @@ struct VoidPlaceholder {
+    virtual void disconnect() = 0;
+};
+
+// HasIsolatedCopy()
```
Wrong comment ? 

All this would fit better in TypeTraits.h

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -957,12 +978,21 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
+        RefPtr<RefCountedSerialFunctionDispatcher> queue;
+        if constexpr (ShouldCallNativePromiseDispatcherGetter<Dispatcher>::value)
+            queue = targetQueue.nativePromiseDispatcher();
+        else if constexpr (IsSmartRef<Dispatcher>::value)
```
Should check if it's a RefCountedSerialFunctionDispatcher and be only passed by reference 

Should use the code prior the introduction of RefCountedSerialFunctionDispatcher instead

jyavenard replied:
As discussed offline, I think we should only ever pass a `RefCountedSerialFunctionDispatcher` and so you call `whenSettled(scope.nativePromiseDispatcher(), ...)` explicitly

jyavenard commented on Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
```diff
@@ -91,24 +91,14 @@ WorkerCacheStorageConnection::~WorkerCacheStorageConnection()
-                downcast<WorkerGlobalScope>(scope).cacheStorageConnection().openCompleted(requestIdentifier, result);
-            }, WorkerRunLoop::defaultMode());
-        });
+    OpenPromise::Producer producer;
```
return invokeAsync(RunLoop::mainthread(), []... { Return mainThreadCinnection->open(...)

No need for a producer

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -942,8 +963,8 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
 public:
-    template<typename SettleFunction>
-    auto whenSettled(RefCountedSerialFunctionDispatcher& targetQueue, SettleFunction&& settleFunction, const Logger::LogSiteIdentifier& callSite = DEFAULT_LOGSITEIDENTIFIER)
+    template<typename Dispatcher, typename SettleFunction>
```
, I don't see why you need this when ScriptExecutionContextNativePromiseDispatcher is a RefCountedSerialFunctionDispatcher

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -955,34 +980,32 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         using LambdaReturnType = decltype(std::declval<LambdaReturnTrait>().template lambda<R1>());
 
-        return whenSettled(targetQueue, [resolveFunction = std::forward<ResolveFunction>(resolveFunction), rejectFunction = std::forward<RejectFunction>(rejectFunction)] (ResultParam result) mutable -> LambdaReturnType {
+        return whenSettled(dispatcher, [resolveFunction = std::forward<ResolveFunction>(resolveFunction), rejectFunction = std::forward<RejectFunction>(rejectFunction)] (ResultParam result) mutable -> LambdaReturnType {
```
`std::forward<Dispatcher>(dispatcher)`

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -993,16 +1016,16 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         using LambdaReturnType = decltype(std::declval<LambdaReturnTrait>().template lambda<R1>());
 
-        return whenSettled(targetQueue, [thisVal = Ref { thisVal }, resolveMethod, rejectMethod] (ResultParam result) -> LambdaReturnType {
+        return whenSettled(dispatcher, [thisVal = Ref { thisVal }, resolveMethod, rejectMethod] (ResultParam result) -> LambdaReturnType {
```
`std::forward<Dispatcher>(dispatcher)`

jyavenard commented on Source/WebCore/dom/ScriptExecutionContextNativePromiseDispatcher.cpp:
```diff
@@ -0,0 +1,74 @@
+    : m_context(context)
+    , m_identifier(context.identifier())
+    , m_source(source)
+    , m_threadId(context.isWorkerGlobalScope() ? Thread::current().uid() : 0)
```
MainThread::id is 1 as using in WorkQueue & RunLoop; if you kept that default it would simplify the isCurrent() implementation.

jyavenard commented on Source/WebCore/workers/WorkerGlobalScope.cpp:
```diff
@@ -170,6 +170,10 @@ void WorkerGlobalScope::prepareForDestruction()
         m_fileSystemStorageConnection->scopeClosed();
+
+    m_nativePromiseCallbacks.forEach([&](auto& callback) {
+        callback.disconnect();
```
on which thread is this called on ?

jyavenard reviewed (commented):
I think there are two conditions we need to cater for:
1- the NativePromise::Request (renamed NativePromiseCallback in the PR) must be called on the required queue.
2- utility like invokeAsync requires to dispatch a callback to then settle the promise.

This solution only covers 1) ; being able to run a callback on the worker thread and guaranteed it will be run is I believe necessary.
Also, any callback capturing a producer while the promise listener is disconnected; the Producer may not be settled; which will cause debug assertions: I doesn't appear that this condition will be satisfied either. So now you have to be very careful in using the NativePromise as you've lost the most important contract of the promise : it *must* be resolved.

I would be more keen to have a way to guarantee that a WorkerThread will be kept alive for as long as a required queue callbacks hasn't been run.

Gecko uses a StrongWorkerRef object, which if held guarantees that the worker thread will not shutdown until the last StrongWorkerRef reference has been released.

If we had such object, we could retrieve the RefCountedGuaranteedSerialFunctionDispatcher (that's a mouthful) from a WorkerThread which would wrap the runnable in a lambda capturing a StrongWorkerRef.

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -273,8 +275,51 @@ class NativePromiseBase : public ThreadSafeRefCounted<NativePromiseBase>  {
 
-template<typename T>
-class NativePromiseRequest;
+class NativePromiseCallback : public ThreadSafeRefCounted<NativePromiseCallback> {
```
The name is not right and is further from what the original name properly implied.

You pass to then/whenSettled a callback, what you get back is a request. It's not a callback.

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1377,46 +1415,6 @@ using GenericPromise = NativePromise<void, void>;
 // A generic, non-exclusive promise type that does the trick for simple use cases.
 using GenericNonExclusivePromise = NativePromise<void, void, PromiseOption::Default | PromiseOption::NonExclusive>;
 
-template<typename PromiseType>
```
can we leave the code in place and just do a rename here?

jyavenard commented on Source/WebCore/dom/ScriptExecutionContext.cpp:
```diff
@@ -851,6 +855,45 @@ CheckedRef<EventLoopTaskGroup> ScriptExecutionContext::checkedEventLoop()
+    bool isCurrent() const final { return m_threadId == Thread::current().uid(); }
+
+    ScriptExecutionContextIdentifier m_identifier;
+    uint32_t m_threadId { 1 };
```
`const uint32_t` ?

youennf replied:
ok

jyavenard commented on Source/WebCore/dom/ScriptExecutionContext.cpp:
```diff
@@ -851,6 +855,45 @@ CheckedRef<EventLoopTaskGroup> ScriptExecutionContext::checkedEventLoop()
+
+RefCountedSerialFunctionDispatcher& ScriptExecutionContext::nativePromiseDispatcher()
+{
+    if (!m_nativePromiseDispatcher)
```
should we add threading / current assertion here?

As this code isn't thread-safe but will be used across-threads

youennf replied:
ScriptExecutionContext::nativePromiseDispatcher is mono thread, it should only be called in context thread.

jyavenard replied:
I think just `dispatcher()` would be better; as it has no direct link to NativePromise as such

jyavenard commented on Source/WebCore/dom/ScriptExecutionContext.h:
```diff
@@ -331,6 +332,27 @@ class ScriptExecutionContext : public SecurityContext, public TimerAlignment {
+                task(WTFMove(result));
+            });
+        });
+        if (weakRequest) {
```
this is rather convoluted. Something seems amiss.

if nativePromiseDispatcher() is a dispatching to a different thread that would cause weakRequest to become null, then the use of WeakPtr isn't thread-safe. 

If it is the same thread, then weakRequest will never be null and the test is unnecessary.

jyavenard commented on Source/WebCore/dom/ScriptExecutionContext.cpp:
```diff
@@ -390,6 +390,10 @@ void ScriptExecutionContext::stopActiveDOMObjects()
     m_deferredPromises.clear();
+
+    m_nativePromiseRequests.forEach([] (auto& request) {
+        request.disconnect();
```
you need to test that the request hasn't been completed as it would otherwise assert.
```suggestion
        if (request)
            request.disconnect();
```

youennf replied:
If request was complete, request would have been destroyed.

jyavenard reviewed (commented):
I like this direction much better, making NativePromiseRequest non-templated is a good move.

I was thinking however that maybe having a `ScriptExecutionListener` with a `NativePromiseRequestForScriptExecution` that would automatically disconnect as needed and on the right thread.

jyavenard commented on Source/WebCore/Modules/cache/CacheStorageConnection.h:
```diff
@@ -29,6 +29,7 @@
 #include "DOMCacheEngine.h"
 #include "RetrieveRecordsOptions.h"
 #include <wtf/HashMap.h>
+#include <wtf/NativePromise.h>
```
```suggestion
#include <wtf/Forward.h>
```
should be sufficient here and include NativePromise.h in the code using it.

youennf replied:
ok

jyavenard commented on Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
```diff
@@ -222,23 +222,26 @@ void DOMCacheStorage::open(const String& name, DOMPromiseDeferred<IDLInterface<D
 void DOMCacheStorage::doOpen(const String& name, DOMPromiseDeferred<IDLInterface<DOMCache>>&& promise)
 {
+    RefPtr context = scriptExecutionContext();
+    if (!context)
```
what happens to the DOMPromise here? does it need to be resolved/rejected?

youennf replied:
Right, it is consistent to reject it.

youennf committed:
Migrate CacheStorageConnection::open to NativePromise
https://bugs.webkit.org/show_bug.cgi?id=266568
rdar://119804320

Reviewed by Jean-Yves Avenard.

The goal is first to simplify CacheStorage operations handling from workers.
We should as a follow-up step allow to send/receive IPC out of main thread, which will be made easier with promises.

Given it runs in workers, we need to have a way to register NativePromise callbacks that are guaranteed to be destroyed in workers.
We do this by registering these callbacks in a WorkerGlobalScope so that we can destroy them when WorkerGlobalScope gets stopped.

To do this, we allow a promise to either take a RefCountedGuaranteedSerialFunctionDispatcher or a NativePromiseDispatcher.
The latter is a RefCountedSerialFunctionDispatcher but without any guarantee that the function will be dispatched or not.
To handle this possibility, the NativePromiseDispatcher is responsible to disconnect the NativePromiseCallback that it is given in registerCallback
in the same context as its RefCountedSerialFunctionDispatcher.

* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/NativePromise.h:
(WTF::NativePromiseRequest::~NativePromiseRequest):
(WTF::NativePromiseRequest::track):
(WTF::NativePromiseRequest::operator bool const):
(WTF::NativePromiseRequest::complete):
(WTF::NativePromiseRequest::disconnect):
* Source/WebCore/Modules/cache/CacheStorageConnection.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::DOMCacheStorage::doOpen):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::open):
(WebCore::WorkerCacheStorageConnection::clearPendingRequests):
(WebCore::WorkerCacheStorageConnection::openCompleted): Deleted.
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::stopActiveDOMObjects):
(WebCore::ScriptExecutionContext::nativePromiseDispatcher):
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::Task::Task):
(WebCore::ScriptExecutionContext::Task::m_isCleanupTask):
(WebCore::ScriptExecutionContext::enqueueTaskWhenPromiseIsSettled):
(WebCore::ScriptExecutionContext::AddConsoleMessageTask::AddConsoleMessageTask):
* Source/WebCore/page/CacheStorageProvider.h:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::open):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/273828@main
